src/dune_tui/dune_tui.ml

```diff
@@ -0,0 +1,133 @@
+  | `Bg_bright_red -> A.(bg lightred)
+  | `Bg_bright_green -> A.(bg lightgreen)
+  | `Bg_bright_yellow -> A.(bg lightyellow)
+  | `Bg_bright_blue -> A.(bg lightblue)
+  | `Bg_bright_magenta -> A.(bg lightmagenta)
+  | `Bg_bright_cyan -> A.(bg lightcyan)
+  | `Bg_bright_white -> A.(bg lightwhite)
+  | `Bold -> A.(st bold)
+  | `Italic -> A.(st italic)
+  | `Dim -> A.(st dim)
+  | `Underline -> A.(st underline)
+
+let attr_of_user_message_style fmt t (pp : User_message.Style.t Pp.t) : unit =
+  let attr =
+    let module A = Notty.A in
+    match (t : User_message.Style.t) with
+    | Loc -> A.(st bold)
+    | Error -> A.(st bold ++ fg red)
+    | Warning -> A.(st bold ++ fg magenta)
+    | Kwd -> A.(st bold ++ fg blue)
+    | Id -> A.(st bold ++ fg yellow)
+    | Prompt -> A.(st bold ++ fg green)
+    | Hint -> A.(st italic ++ fg white)
+    | Details -> A.(st dim ++ fg white)
+    | Ok -> A.(st italic ++ fg green)
+    | Debug -> A.(st underline ++ fg lightcyan)
+    | Success -> A.(st bold ++ fg green)
+    | Ansi_styles l ->
+      List.fold_left ~init:A.empty l ~f:(fun attr s ->
+          A.(attr ++ attr_of_ansi_color_style s))
+  in
+  Notty.I.pp_attr attr Pp.to_fmt fmt pp
+
+let image_of_user_message_style_pp =
+  Notty.I.strf "%a@."
+    (Pp.to_fmt_with_tags ~tag_handler:attr_of_user_message_style)
+
+module Tui () = struct
+  module Term = Notty_unix.Term
+
+  let term = Term.create ~nosig:false ()
+
+  let start () = Unix.set_nonblock Unix.stdin
+
+  let image ~status_line ~messages =
+    let status =
+      match (status_line : User_message.Style.t Pp.t option) with
+      | None -> []
+      | Some message -> [ image_of_user_message_style_pp message ]
+    in
+    let messages =
+      List.map messages ~f:(fun msg ->
+          image_of_user_message_style_pp (User_message.pp msg))
+    in
+    Notty.I.vcat (messages @ status)
+
+  let render (state : Dune_console.Threaded.state) =
+    let messages = Queue.to_list state.messages in
+    let image = image ~status_line:state.status_line ~messages in
+    Term.image term image
+
+  let rec handle_user_events ~now ~time_budget _mutex =
+    (* We check for any user input and handle it. If we go over the
+       [time_budget] we give up and continue. *)
+    let input_fds, _, _ = Unix.select [ Unix.stdin ] [] [] time_budget in
+    match input_fds with
+    | [] ->
+      now +. time_budget
+      (* Nothing to do, we return the time at the end of the time budget. *)
+    | _ :: _ -> (
+      (* TODO if anything fancy is done in the UI in the future we need to lock
+         the state with the provided mutex *)
+      match Term.event term with
+      | `Key (`ASCII 'q', _) ->
+        (* When we encounter q we make sure to quit by signaling termination. *)
+        Unix.kill (Unix.getpid ()) Sys.sigterm;
+        (* We return either the intended end time or the current time if we
+           went over it. *)
+        Float.max (now +. time_budget) (Unix.gettimeofday ())
```

Comment:
This function should always return the most accurate `$now` at the end of execution. So there's no need for the max here. It should just be `Unix.gettimeofday ()`.